design/SOURCE&DESTINATION.md

```diff
@@ -0,0 +1,222 @@
+
+    abstract fun read(): Buffer
+
+    abstract suspend fun awaitBytes(): Boolean
+
+    abstract fun cancel(cause: Throwable = CancellationException("Source has been cancelled"))
+}
+```
+
+## Destination
+
+To handle different outputs there is another abstraction: `Destination`. It is responsible for abstracting different
+outputs.
+
+To write data to the destination we can use `write` method:
+
+```kotlin
+val destination = mySocket.destination
+
+val count = destination.write(buffer)
+println("$count bytes written")
+```
+
+Write will write as much data(between `readPosition` and `writePosition`) it can to the destination. It will
+modify `readPosition` and return how many bytes were written.
+
+While the `Source` is responsible for creating `Buffer`, `Destination` is responsible for releasing it: if the `buffer`
+has no bytes available after write, it will be automatically released.
+
+How could you write a whole `Buffer` to the IO? The `spin` on the `write` method is not a good idea, so there is a
+method `awaitFreeSpace()` that will suspend until the destination has some free space to write:
+
+```kotlin
+fun Destination.writeBuffer(buffer: Buffer) {
+    while (true) {
+        write(buffer)
+
+        if (buffer.isEmpty()) return
+        awaitFreeSpace()
+    }
+}
+```
+
+### How to end the conversation?
+
+With `close()` method It will terminate `Destination` and release all resources. If there are some pending bytes they
+will be sent:
+
+```kotlin
+fun Destination.writeTrailer(trailer: ByteArray) {
+    writeFully(trailer)
+    close()
+}
+```
+
+It also can be useful to save some additional information about why and where `Destination` was closed, so you can pass
+`cause` to the close method. There is also `isClosed` flag to check if the destination is already closed.
+
+### Flush
+
+It's hard to write human-readable justification of having flush method without introducing a new abstractions, so I will
+keep it as TODO("Rewrite this").
+
+Not human-readable justification:
+
+If you have chained transformations on the `Destination` like:
+
+```kotlin
+val destination = mySocket.destination
+    .compressed()
+    .buffered()
+```
+
+and a single `message`. There is a common pattern when you want to send this message and make sure that all data is
+actually written to file or socket(for instance it can be a http message header). Some transformations could keep an
+internal buffer because of efficiency reasons.
+
+To make sure that all data is written to the destination, you can call `flush()`. Flush is a suspend operation which
+will suspend until all data is actually written to the final destination:
+
+```kotlin
+destination.writeFully(message)
+destination.flush() 
+// At this point message should be compressed and written to the socket.
+```
+
+## Final Listing
+
+```kotlin
+abstract class Destination {
+    abstract val isClosed: Boolean
+    abstract val closeCause: Throwable?
+
+    abstract fun write(buffer: Buffer): Int
```

Comment:
I like `write` because of simplicity and mirroring of `read`. But indeed, it can be confusing. `writeAsMuchAsPossible` is better than `writeAvailable`, but looks ugly and still can be misunderstood.